### PR TITLE
create new lexical env inside switch statement blocks, fixes #T7324

### PR DESCRIPTION
**Pull request by jayphelps**
As described in https://phabricator.babeljs.io/T7324, this would have been a lot easier I think if the AST supported [`CaseBlock` from the spec](http://www.ecma-international.org/ecma-262/6.0/#sec-switch-statement), so instead we sort of have to fudge that `SwitchStatement` is a block.

It's also possible I'm totally missing a much cleaner solution.

### CODE REVIEW

File: packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/actual.js [hunk 1]
```diff
@@ -9,3 +11,15 @@ switch (a) {
...
+  case true:
+    let a = false;
+    let b = true;
+    let c = 4;
+    let d = 5;
+
+  case false:
+    class e {}
+    function f() {}
+}
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/actual.js]: I'm not familiar with this code, but does this cover cases like

```
    switch(true){
      case true:
        let a = 4;
      case false:
        a = 5;
        console.log(a);
    }
```
properly? We should make sure the tests cover that to be sure.

File: packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js [hunk 2]
```diff
@@ -9,3 +11,17 @@ switch (a) {
...
+
+switch (true) {
+  case true:
+    var _a = false;
+    var _b = true;
+    var _c = 4;
+    var _d = 5;
+
+  case false:
+    class _e {}
```
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js]: Interesting that this transform doesn't convert that to `var _e = class e {}`, seems like a separate oversight.

File: packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/actual.js (hunk 1 above)
Comment (jayphelps) [inline packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/actual.js]: It doesn't currently, I'll add it. 👍

to be honest, there are more edge cases not currently covered with lexical scope in general but IMO that's outside the scope of this PR since I'm just trying to fix this bug 😏

Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/actual.js]: Yeah, understandable.

File: packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js (hunk 2 above)
Comment (jayphelps) [inline packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js]: hmm why would it? it doesn't have `transform-class` enabled so AFAIK it should just pass syntax straight through as-is. Also, those two are semantically the same aren't they?

File: packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js [hunk 3]
```diff
@@ -1,11 +1,29 @@
...
+function f() {}
 
 switch (a) {
   case true:
     var c = 2;
+    var foo = true;
     break;
   case false:
     var d = 3;
+    foo = false;
```
Comment (jayphelps) [inline packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js]: @loganfsmyth ^^

File: packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js (hunk 2 above)
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js]: Class declarations are block-scoped just like `let`/`const`.

     class e {}

is essentially 

    let e = class {}

but looks like it doesn't take that into account.

Comment (jayphelps) [inline packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js]: Right, but why would it need to transform it to `ClassExpression`? it underscored it (i.e. performed hygiene) correctly, `class _e {}` so they no longer collide at runtime, just like all the others.

Comment (jayphelps) [inline packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js]: e.g. they to show they aren't the same, even though the transpiler might today handle them the same if they have the same idenifiers

image

Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js]: I'm just being nit-picky :P. If the objective of this transform is to process block scoping, it surprised me to see a structure left that is inherently block scoped, like a class declaration. I'm sure it was left on purpose since there's no actual reason to bother converting it here, since there isn't going to be an environment that has implemented classes but not implemented block scoping. Just commenting that it's interesting, this is what I get for doing code review past midnight.

Comment (jayphelps) [inline packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js]: I hope I didn't offend you, I'm just debating 😄

**Comment by jmm** [comment]
@jayphelps FYI, Babel since v6 uses [its own AST](https://github.com/babel/babel/commit/07f0bd4ea93657f8b3158f6aedd8e11721d88015) that diverges from ESTree.

File: packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js (hunk 2 above)
Comment (loganfsmyth) [inline packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/switch/expected.js]: Not offended, I promise haha :) If the internal and external names are different, those will behave differently, but if they are the same, I can't think of a case where they'd have an detectable difference.

**Comment by jayphelps** [comment]
I have no idea how code coverage was reduced. I didn't touch any of the files it lists?

**Comment by hzoo** [comment]
I'm not sure what haven't to our code coverage recently because it has been saying 100% which doesn't make sense to me.

**Comment by jayphelps** [comment]
Let me know if I need to do something to fix the code coverage. Otherwise I'll wait for merge or additional comments.

**Comment by jayphelps** [comment]
What's the word on this?

**Comment by loganfsmyth** [comment]
Okay, I think this solution seems right, and we're going to learn get anywhere if we just leave this hanging, so I'm going to merge. We'll just want to keep an eye out for possible breakage.
